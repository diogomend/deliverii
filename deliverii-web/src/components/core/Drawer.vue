<template>
  <v-navigation-drawer
    id="core-navigation-drawer"
    v-model="showDrawer"
    :expand-on-hover="expandOnHover"
    :right="$vuetify.rtl"
    mobile-break-point="960"
    app
    width="260"
    v-bind="$attrs"
  >
    <v-divider class="mb-1" />

    <v-list dense nav>
      <v-list-item-group active-class="text--accent-4">
        <v-list-item v-for="(page, index) in items" :key="index" :to="page.to">
          <v-list-item-icon>
            <v-icon>{{ page.icon }}</v-icon>
          </v-list-item-icon>
          <v-list-item-title>{{ page.title }}</v-list-item-title>
        </v-list-item>
      </v-list-item-group>
    </v-list>

    <v-divider class="mb-2" />
  </v-navigation-drawer>
</template>

<script>
import { mapGetters, mapActions } from "vuex";
import MobileMixin from "@/mixins/mobileMixin";
import { ADMIN_DRAWER, USER_DRAWER } from "@/config/config";
export default {
  name: "DashboardCoreDrawer",

  props: {
    expandOnHover: {
      type: Boolean,
      default: false
    }
  },
  mixins: [MobileMixin],

  /*data: () => ({
    items: [
      {
        icon: "mdi-home",
        title: "Home",
        to: "/admin/"
      },
      {
        icon: "mdi-store",
        title: "Restaurants",
        to: "/admin/restaurants"
      },
      {
        icon: "mdi-food-variant",
        title: "Orders",
        to: "/admin/orders"
      },
    ]
  }),*/

  computed: {
    ...mapGetters({
      drawer: "appState/drawer",
      user: "auth/getUser"
    }),
    computedItems() {
      return this.items.map(this.mapItem);
    },
    items() {
      if (this.user.isManager) {
        return ADMIN_DRAWER;
      }

      return USER_DRAWER;
    },
    profile() {
      return {
        avatar: true,
        title: "avatar"
      };
    },
    showDrawer: {
      get() {
        return !this.isMobile || this.drawer;
      },
      set(value) {
        this.toggleDrawer(value);
      }
    }
  },

  methods: {
    mapItem(item) {
      return {
        ...item,
        children: item.children ? item.children.map(this.mapItem) : undefined,
        title: item.title
      };
    },
    ...mapActions("appState", ["toggleDrawer"])
  }
};
</script>

<style lang="sass">
@import '~vuetify/src/styles/tools/_rtl.sass'

#core-navigation-drawer
  .v-list-group__header.v-list-item--active:before
    opacity: .24

  .v-list-item
    &__icon--text,
    &__icon:first-child
      justify-content: center
      text-align: center
      width: 20px

      +ltr()
        margin-right: 24px
        margin-left: 12px !important

      +rtl()
        margin-left: 24px
        margin-right: 12px !important

  .v-list--dense
    .v-list-item
      &__icon--text,
      &__icon:first-child
        margin-top: 10px

  .v-list-group--sub-group
    .v-list-item
      +ltr()
        padding-left: 8px

      +rtl()
        padding-right: 8px

    .v-list-group__header
      +ltr()
        padding-right: 0

      +rtl()
        padding-right: 0

      .v-list-item__icon--text
        margin-top: 19px
        order: 0

      .v-list-group__header__prepend-icon
        order: 2

        +ltr()
          margin-right: 8px

        +rtl()
          margin-left: 8px
</style>
