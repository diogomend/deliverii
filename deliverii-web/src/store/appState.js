const getTimeout = alert => {
  return "timeout" in alert ? alert.timeout : 5000;
};

const actions = {
  toggleDrawer: ({ commit }, toggle) => {
    commit("TOGGLE_DRAWER", toggle);
  },
  toggleLoading: ({ commit }, toggle) => {
    commit("TOGGLE_LOADING", toggle);
  },
  addAlert: ({ commit, getters }, alert) => {
    const timeout = getTimeout(alert);
    const alerts = getters.getAlerts;
    const repeatedIndex = alerts.findIndex(val => {
      return val.type === alert.type && val.message === alert.message;
    });
    let id;

    do {
      id = Date.now() + Math.random();
    } while (getters.getAlerts.some(elem => elem.id === id));

    commit("ADD_ALERT", { ...alert, id });

    if (repeatedIndex !== -1) {
      commit("REMOVE_ALERT", repeatedIndex);
    }

    setTimeout(() => {
      const index = getters.getAlerts.findIndex(elem => elem.id === id);

      if (index !== -1) {
        commit("REMOVE_ALERT", index);
      }
    }, timeout);
  }
};

const mutations = {
  ADD_ALERT(state, { type, message, id }) {
    state.alert.push({
      type,
      message,
      id
    });
  },
  REMOVE_ALERT(state, index) {
    state.alert.splice(index, 1);
  },
  TOGGLE_LOADING(state, loading) {
    state.isLoadingSession = loading;
  },
  TOGGLE_DRAWER(state, drawer) {
    state.drawer = drawer;
  }
};

const getters = {
  getAlerts: state => state.alert,
  getLoadingSession: state => state.isLoadingSession,
  drawer: state => state.drawer
};

const state = () => ({
  drawer: false,
  isLoadingSession: true,
  alert: []
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
