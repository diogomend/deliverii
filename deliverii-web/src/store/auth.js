import auth from "../service/auth";
const actions = {
  login: async ({ commit, dispatch }, { email, password }) => {
    const res = await auth.login(email, password);
    if (res) {
      commit("SET_USER", res.user);
      commit("SET_TOKEN", res.token);
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Success`
        },
        { root: true }
      );
      return res;
    }
    return false;
  },
  register: async ({ dispatch }, registerObj)=> {
    const res = await auth.register(registerObj);
    if (res) {
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Registered Successfully`
        },
        { root: true }
      );
      return res;
    }
    return false;
  },
  logout({ commit }) {
    commit("SET_USER", false);
    commit("SET_TOKEN", false);

    auth.logoutUser();
  },

  async loadSession({ commit }) {
    const isLoggedIn = auth.isLoggedIn();
    if (isLoggedIn) {
      const user = auth.getUserInfo();
      commit("SET_USER", user);
    }
  },
};

const mutations = {
  SET_USER(state, user) {
    state.user = user;
  },
  SET_TOKEN(state, token) {
    state.token = token;
  }
};

const getters = {
  getUser: state => state.user,
  isLoggedIn: state => !!state.user
};

const state = () => ({
  user: false,
  token: false
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
