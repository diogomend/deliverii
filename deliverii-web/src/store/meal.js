import meal from "../service/meal";

const actions = {
  fetchMeals: async (_, restaurantID) => {
    return await meal.fetchMeals(restaurantID);
  },
  fetchMeal: async (_, { restaurant, meal: mealID }) => {
    const mealList = await meal.fetchMeal(restaurant, mealID);
    if (mealList && mealList.length) {
      return mealList[0];
    }

    return false;
  },

  updateMeal: async ({ dispatch }, { restaurant, mealID, mealObj }) => {
    const res = await meal.updateMeal(restaurant, mealID, mealObj);
    if (res) {
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Success`
        },
        { root: true }
      );

      return true;
    }

    return false;
  },

  createMeal: async ({ dispatch }, { restaurant, mealObj }) => {
    const res = await meal.createMeal(restaurant, mealObj);
    if (res) {
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Success`
        },
        { root: true }
      );

      return true;
    }

    return false;
  }
};

const mutations = {};

const getters = {};

const state = () => ({
  meals: []
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
