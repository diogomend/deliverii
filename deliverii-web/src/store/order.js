import order from "../service/order";

const actions = {
  fetchAll: async ({ commit }, id) => {
    const res = await order.fetchAll(id);
    commit("ADD_ORDERS", res);
  },
  fetchOrder: async (_, id) => {
    return await order.fetchOrder(id);
  },
  updateStatus: async (_, { orderID, status }) => {
    return await order.updateStatus(orderID, status);
  },
  loadCart: async ({ commit }) => {
    const cart = await order.loadCart();
    if (cart) {
      commit("ADD_CART", JSON.parse(cart));
    }
  },
  addCart({ commit }, cart) {
    order.saveCart(cart);
    commit("ADD_CART", cart);
  },
  addAddress({ commit }, address) {
    commit("ADD_ADDRESS", address);
  },
  async makeOrder({ dispatch, getters }) {
    const cartObj = getters.getCart;
    const meals = cartObj.cart.map(item => {
      return {
        meal: item._id,
        quantity: item.quantity
      };
    });

    const request = {
      meals: meals,
      restaurantID: cartObj.restaurantID,
      address: cartObj.address
    };

    const ret = await order.makeOrder(request);
    if (ret._id) {
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Order created successfully`
        },
        { root: true }
      );
      return ret._id;
    }

    return false;
  }
};

const mutations = {
  ADD_ORDERS(state, orders) {
    state.orders = orders;
  },
  ADD_CART(state, cart) {
    state.cart = cart;
  },
  ADD_ADDRESS(state, address) {
    state.cart.address = address;
  }
};

const getters = {
  getOrders: state => state.orders,
  getCart: state => state.cart
};

const state = () => ({
  orders: [],
  cart: {}
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
