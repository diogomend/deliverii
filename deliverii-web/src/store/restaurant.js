import restaurant from "../service/restaurant";

const actions = {
  fetchAll: async ({ commit }) => {
    const res = await restaurant.fetchAll();
    commit("ADD_RESTAURANTS", res);
  },
  fetchRestaurantInfo: async (_, id) => {
    return await restaurant.fetchRestaurant(id);
  },
  deleteRestaurant: async ({ dispatch }, id) => {
    const res = await restaurant.deleteRestaurant(id);
    if (res) {
      await dispatch(
        "appState/addAlert",
        { type: "success", message: `Delete successfully` },
        { root: true }
      );
    }

    return res;
  },
  createRestaurant: async ({ dispatch }, restaurantObj) => {
    const res = await restaurant.createRestaurant(restaurantObj);
    if (res) {
      await dispatch(
        "appState/addAlert",
        { type: "success", message: `Created successfully` },
        { root: true }
      );
    }

    return res;
  },

  addBlacklist: async ({ dispatch }, { restaurantID, customerID }) => {
    const res = await restaurant.addBlacklist(restaurantID, customerID);
    if (res) {
      await dispatch(
        "appState/addAlert",
        { type: "success", message: `Added to blacklist` },
        { root: true }
      );
    }

    return res;
  }
};

const mutations = {
  ADD_RESTAURANTS(state, restaurants) {
    state.restaurants = restaurants;
  }
};

const getters = {
  getRestaurants: state => state.restaurants
};

const state = () => ({
  restaurants: []
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
