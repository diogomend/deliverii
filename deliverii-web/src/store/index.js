import Vue from "vue";
import Vuex from "vuex";
import appState from "@/store/appState";
import auth from "@/store/auth";
import order from "@/store/order";
import restaurant from "@/store/restaurant";
import meal from "@/store/meal";

Vue.use(Vuex);

export const storeOptions = {
  modules: {
    appState,
    auth,
    order,
    restaurant,
    meal
  }
};

export default new Vuex.Store(storeOptions);
