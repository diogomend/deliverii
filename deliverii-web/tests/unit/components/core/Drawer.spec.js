import Vuetify from 'vuetify';
import Vue from 'vue';
import Vuex from "vuex";
import Drawer from "@/components/core/Drawer.vue";
import { shallowMount } from "@vue/test-utils"
import { storeOptions } from "@/store/index"

Vue.use(Vuex);
Vue.use(Vuetify);


describe("Drawer.vue", () => {
    let store;

    beforeEach(() => {
        jest.clearAllMocks();
        store = new Vuex.Store(storeOptions);
      });

      it("is a Vue instance", () => {
        const wrapper = shallowMount(Drawer, {
          store,
        });
    
        const template = wrapper.html();
        expect(template).toMatchSnapshot();
      });

      it("should get manager drawer", () => {
        storeOptions.modules.auth.getters.getUser = jest.fn().mockImplementation(() => {
            return {isManager: true}
        });
        store = new Vuex.Store(storeOptions);
        const wrapper = shallowMount(Drawer, {
          store,
          computed: {

          }
        });
    
        expect(wrapper.vm.items.length).toEqual(3);
      });

      it("should get user drawer", () => {
        storeOptions.modules.auth.getters.getUser = jest.fn().mockImplementation(() => {
            return {isManager: false}
        });
        store = new Vuex.Store(storeOptions);
        const wrapper = shallowMount(Drawer, {
          store,
          computed: {

          }
        });
    
        expect(wrapper.vm.items.length).toEqual(2);
        expect(wrapper.vm.computedItems.length).toEqual(2);
      });

      it("should call toggleDrawer on setting", () => {
        const mockFn = jest.fn();
        storeOptions.modules.appState.actions.toggleDrawer = mockFn;
        store = new Vuex.Store(storeOptions);
        const wrapper = shallowMount(Drawer, {
          store,
          computed: {
            isMobile: true
          }
        });
    
        wrapper.vm.showDrawer = true;
        expect(mockFn).toHaveBeenCalled();
        expect(wrapper.vm.showDrawer).toBeTruthy();
      });
})