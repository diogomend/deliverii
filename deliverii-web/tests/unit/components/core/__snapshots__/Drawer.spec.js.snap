// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer.vue is a Vue instance 1`] = `
<v-navigation-drawer-stub app="true" height="100vh" minivariantwidth="56" mobilebreakpoint="960" src="" tag="nav" width="260" value="true" id="core-navigation-drawer">
  <v-divider-stub class="mb-1"></v-divider-stub>
  <v-list-stub tag="div" tile="true" dense="true" nav="true">
    <v-list-item-group-stub activeclass="text--accent-4">
      <v-list-item-stub activeclass="" to="/" tag="div">
        <v-list-item-icon-stub>
          <v-icon-stub>mdi-store</v-icon-stub>
        </v-list-item-icon-stub>
        <v-list-item-title-stub>Restaurants</v-list-item-title-stub>
      </v-list-item-stub>
      <v-list-item-stub activeclass="" to="/orders" tag="div">
        <v-list-item-icon-stub>
          <v-icon-stub>mdi-food-variant</v-icon-stub>
        </v-list-item-icon-stub>
        <v-list-item-title-stub>Orders</v-list-item-title-stub>
      </v-list-item-stub>
    </v-list-item-group-stub>
  </v-list-stub>
  <v-divider-stub class="mb-2"></v-divider-stub>
</v-navigation-drawer-stub>
`;
